File: openEarsPlugin.js
/**
 * Phonegap OpenEars Plugin
 */
var OpenEarsPlugin = function() {};

OpenEarsPlugin.prototype.callbacks = {};
OpenEarsPlugin.hypothesis = {};

OpenEarsPlugin.prototype.isValidHypothesis = function(hypothesis){
    return hypothesis != '' && hypothesis != '(null)';
}

OpenEarsPlugin.prototype.setCallback = function(name, fn){
    OpenEarsPlugin.prototype.callbacks[name] = fn;
}

OpenEarsPlugin.prototype.log = function(msg){
    $('#log').append('<h1>'+msg+'</h1>');
    PhoneGap.exec('openEarsPlugin.log',msg);
};

OpenEarsPlugin.prototype.startAudioSession = function(){
    PhoneGap.exec('openEarsPlugin.startAudioSession');
};

OpenEarsPlugin.prototype.startListeningWithLanguageModelAtPath = function(options){
    PhoneGap.exec('openEarsPlugin.pocketsphinxControllerStartListeningWithLanguageModelAtPath',options.languagemodel, options.dictionary, options);
};

OpenEarsPlugin.prototype.changeLanguageModelToFile = function(options){
    PhoneGap.exec('openEarsPlugin.pocketsphinxControllerChangeLanguageModelToFile', options.languagemodel, options.dictionary);
}

OpenEarsPlugin.prototype.generateLanguageModel = function(languageArrayCSV){
    PhoneGap.exec('openEarsPlugin.languageModelGeneratorGenerateLanguageModelFromArray',languageArrayCSV);
}


OpenEarsPlugin.prototype.stopListening = function(){
    PhoneGap.exec('openEarsPlugin.pocketsphinxControllerStopListening');
};

OpenEarsPlugin.prototype.suspendRecognition = function(){
    PhoneGap.exec('openEarsPlugin.pocketsphinxControllerSuspendRecognition');
};

OpenEarsPlugin.prototype.resumeRecognition = function(){
    PhoneGap.exec('openEarsPlugin.pocketsphinxControllerResumeRecognition');
};

OpenEarsPlugin.prototype.say = function(phrase){
    PhoneGap.exec('openEarsPlugin.fliteControllerSay',phrase);
};


PhoneGap.addConstructor(function() {
    if(!window.plugins){window.plugins = {};}
    window.plugins.openEarsPlugin = new OpenEarsPlugin();
    
    // Did start listening
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidStartListening', function(){
        $(document).trigger('didstartlistening');
    });

    // Did stop listening
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidStopListening', function(){
        $(document).trigger('didstoplistening');
    });

    // Did detect speech
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidDetectSpeech', function(){
	$(document).trigger('diddetectspeech');
    });

    // Did detect finished speech
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidDetectFinishedSpeech', function(){
	$(document).trigger('diddetectfinishedspeech');
    });

    // Did change language model
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidChangeLanguageModelToFile', function(){
	$(document).trigger('didchangelanguagemodel');
    });

    // Did receive hypothesis
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidReceiveHypothesis', function(hypothesis, recognitionScore, utteranceID){
	window.plugins.openEarsPlugin.hypothesis = {
                hypothesis: hypothesis
            ,   recognitionScore: recognitionScore
            ,   utteranceID: utteranceID
        }
        $(document).trigger('didreceivehypothesis');
    });

    // Did suspend recognition
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidSuspendRecognition', function(){
	$(document).trigger('didsuspendrecognition');
    });

    // Did resume recognition
    window.plugins.openEarsPlugin.setCallback('pocketsphinxDidResumeRecognition', function(){
	$(document).trigger('didresumerecognition');
    });
    
    
});

File: openEarsPlugin.m
//
//  openEarsPlugin.mm
//  rhymble
//
//  Created by Richard Telep on 6/25/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "openEarsPlugin.h"

@implementation openEarsPlugin


/*
 *  ++++++++++++++++++++++++++++++++++++++++
 *  Fields
 *  ++++++++++++++++++++++++++++++++++++++++
 */  
@synthesize audio_session_manager;
-(AudioSessionManager *) audio_session_manager{
    if (audio_session_manager == nil){
        audio_session_manager = [[AudioSessionManager alloc] init];
    }
    return audio_session_manager;
}

@synthesize pocket_sphinx_controller;
-(PocketsphinxController *) pocket_sphinx_controller{
    if (pocket_sphinx_controller == nil){
        pocket_sphinx_controller = [[PocketsphinxController alloc] init];
    }
    return pocket_sphinx_controller;
}

@synthesize openears_events_observer;
-(OpenEarsEventsObserver *) openears_events_observer{
    if (openears_events_observer == nil){
        openears_events_observer = [[OpenEarsEventsObserver alloc] init];
    }
    return openears_events_observer;
}

@synthesize flite_controller;
-(FliteController *) flite_controller{
    if (flite_controller == nil){
        flite_controller = [[FliteController alloc] init];
    }
    return flite_controller;
}

@synthesize current_language_model;
@synthesize current_dictionary;

/*
 *  ++++++++++++++++++++++++++++++++++++++++
 *  AudioSessionManager methods
 *  ++++++++++++++++++++++++++++++++++++++++
 */  


-(void)startAudioSession:(NSArray *)arguments withDict:(NSDictionary *)options{
    
    // Start AudioSessionManager
    [self.audio_session_manager startAudioSession];

    // This class will be the delegate of the OpenEarsEventsObserver object.
    [self.openears_events_observer setDelegate:self];


}

/*
 *  ++++++++++++++++++++++++++++++++++++++++
 *  Log
 *  ++++++++++++++++++++++++++++++++++++++++
 */

-(void)log:(NSString*)msg withDict:(NSMutableDictionary *)options{
    NSLog(@"%@",msg);
}


/*  
 *  ++++++++++++++++++++++++++++++++++++++++
 *  PocketsphinxController methods
 *  ++++++++++++++++++++++++++++++++++++++++
 */

-(void)pocketsphinxControllerStartListeningWithLanguageModelAtPath:(NSArray *)arguments withDict:(NSDictionary *)options{
    self.current_language_model = [NSString stringWithFormat:@"%@/%@",[[NSBundle mainBundle] resourcePath], [arguments objectAtIndex:0]];
    self.current_dictionary = [NSString stringWithFormat:@"%@/%@",[[NSBundle mainBundle] resourcePath], [arguments objectAtIndex:1]];
    [self.pocket_sphinx_controller startListeningWithLanguageModelAtPath:self.current_language_model dictionaryAtPath:self.current_dictionary languageModelIsJSGF:NO];
}

-(void)pocketsphinxControllerChangeLanguageModelToFile:(NSArray *) arguments withDict:(NSDictionary *)options{
    self.current_language_model = [NSString stringWithFormat:@"%@/%@",[[NSBundle mainBundle] resourcePath], [arguments objectAtIndex:0]];
    self.current_dictionary = [NSString stringWithFormat:@"%@/%@",[[NSBundle mainBundle] resourcePath], [arguments objectAtIndex:1]];
    [self.pocket_sphinx_controller changeLanguageModelToFile:self.current_language_model withDictionary:self.current_dictionary];
}

-(void)pocketsphinxControllerStopListening:(NSArray *)arguments withDict:(NSDictionary *)options{
    [self.pocket_sphinx_controller stopListening];
}

-(void)pocketsphinxControllerSuspendRecognition:(NSArray *)arguments withDict:(NSDictionary *)options{
    [self.pocket_sphinx_controller suspendRecognition];
}

-(void)pocketsphinxControllerResumeRecognition:(NSArray *)arguments withDict:(NSDictionary *)options{
    [self.pocket_sphinx_controller resumeRecognition];
}

/*
 *  ++++++++++++++++++++++++++++++++++++++++
 *  FliteController methods
 *  ++++++++++++++++++++++++++++++++++++++++
 */

-(void)fliteControllerSay:(NSString *)phrase withDict:(NSDictionary *)options{
    NSString* phrase_out = [[NSString alloc] initWithFormat:@"%@",phrase];
    NSLog(phrase_out);
    [self.flite_controller say:phrase_out withVoice:@"cmu_us_slt"];
    [phrase_out release];
}

/*
 *  ++++++++++++++++++++++++++++++++++++++++
 *  OpenEarsEventsObserver delegate methods
 *  ++++++++++++++++++++++++++++++++++++++++
 */

- (void) pocketsphinxDidReceiveHypothesis:(NSString *)hypothesis recognitionScore:(NSString *)recognitionScore utteranceID:(NSString *)utteranceID {
    NSLog(@"Pocketsphinx received a hypothesis is %@ with a score of %@ and an ID of %@", hypothesis, recognitionScore, utteranceID);
    NSString* jsString = [[NSString alloc] initWithFormat:@"window.plugins.openEarsPlugin.callbacks.pocketsphinxDidReceiveHypothesis(\"%@\",%@,%@);",hypothesis,recognitionScore,utteranceID];
    [self writeJavascript:jsString];
	[jsString release];
}

// This doesn't appear to be working:
//- (void) pocketsphinxDidChangeLanguageModelToFile:(NSString *)newLanguageModelPathAsString andDictionary:(NSString *)newDictionaryPathAsString {
//	NSLog(@"Pocketsphinx is now using the following language model: \n%@ and the following dictionary: %@",newLanguageModelPathAsString,newDictionaryPathAsString);
//    NSString* jsString = [[NSString alloc] initWithFormat:@"window.plugins.openEarsPlugin.callbacks.pocketsphinxDidChangeLanguageModelToFile(\"%@\");",newDictionaryPathAsString];
//    NSLog(jsString);
//    [self writeJavascript:jsString];
//	[jsString release];
//}

        // There are numerous other methods to implement here.


/*
 *  ++++++++++++++++++++++++++++++++++++++++
 *  Cleanup
 *  ++++++++++++++++++++++++++++++++++++++++
 */

-(void) dealloc {
    [audio_session_manager release];
    [pocket_sphinx_controller release];
    openears_events_observer.delegate = nil;
    [openears_events_observer release];
    [current_language_model release];
    [current_dictionary release];
    [super dealloc];
}




@end
